frontend/appflowy_flutter/lib/plugins/database_view/grid/presentation/grid_page.dart

```diff
@@ -216,67 +201,102 @@ class _GridRowsState extends State<_GridRows> {
-      listener: (context, state) {
-        state.reason.whenOrNull(
-          insert: (item) {
-            _key.currentState?.insertItem(item.index);
-          },
-          delete: (item) {
-            _key.currentState?.removeItem(
-              item.index,
-              (context, animation) =>
-                  _renderRow(context, item.rowInfo, animation),
+    return BlocBuilder<GridFilterMenuBloc, GridFilterMenuState>(
+      builder: (context, filterState) {
+        return BlocBuilder<SortMenuBloc, SortMenuState>(
+          builder: (context, sortState) {
+            return BlocConsumer<GridBloc, GridState>(
+              listenWhen: (previous, current) =>
+                  previous.reason != current.reason,
+              listener: (context, state) {
+                state.reason.whenOrNull(
+                  insert: (item) {
+                    _key.currentState?.insertItem(item.index);
+                  },
+                  delete: (item) {
+                    _key.currentState?.removeItem(
+                      item.index,
+                      (context, animation) => _renderRow(
+                        context,
+                        item.rowInfo,
+                        animation: animation,
+                      ),
+                    );
+                  },
+                );
+              },
+              buildWhen: (previous, current) {
+                return current.reason.maybeWhen(
+                  reorderRows: () => true,
+                  reorderSingleRow: (reorderRow, rowInfo) => true,
+                  delete: (item) => true,
+                  insert: (item) => true,
+                  orElse: () => false,
+                );
+              },
+              builder: (context, state) {
+                final rowInfos = context.watch<GridBloc>().state.rowInfos;
+
+                return SliverToBoxAdapter(
+                  child: ReorderableListView.builder(
+                    key: _key,
+                    buildDefaultDragHandles: false,
+                    proxyDecorator: (child, index, animation) => Material(
+                      color: Colors.white.withOpacity(.1),
+                      child: Opacity(
+                        opacity: .5,
+                        child: child,
+                      ),
+                    ),
+                    shrinkWrap: true,
+                    onReorder: (fromIndex, newIndex) {
+                      final toIndex =
+                          newIndex > fromIndex ? newIndex - 1 : newIndex;
+
+                      context
+                          .read<GridBloc>()
+                          .add(GridEvent.moveRow(fromIndex, toIndex));
+                    },
+                    itemCount: rowInfos.length,
+                    itemBuilder: (BuildContext context, int index) {
+                      final RowInfo rowInfo = rowInfos[index];
+                      if (index >= rowInfos.length) {
```

Comment:
will the `index >= rowInfos.length`? maybe we can add an assert here.

Reply:
Actually it wouldn't make sense, I am not sure why this was here before, removed.
(itemCount is rowInfos.length, so if there is no items then it doesn't build one.)